### PR TITLE
lib/beacon, lib/discover: Send IPv4 limited broadcast when address listing fails (fixes #1628)

### LINKED ISSUE #1628: "No availabile source device" - sync is stuck
Me and a friend are sharing a folder. It worked fine for two weeks, now one file is stuck.
On my side the GUI says it's syncing 77% and that's not changing. No error messages.
The file is on my machine and the remote machine is unable to fetch it for some reason.

I have experienced numerous system crashes recently, maybe that could have caused the trouble?
I can remove that file, but it doesn't seem like a solution to me.

Here's a log from the friend's machine (I removed IPs, names and syncthing IDs):

	[monitor] 16:37:17 INFO: Starting syncthing
	[OCBUJ] 16:37:18 INFO: syncthing v0.10.30 (go1.4.2 linux-amd64 default) unknown-user@syncthing-builder 2015-03-29 07:46:44 UTC
	[OCBUJ] 16:37:18 INFO: My ID: OCBUJGQ-[sensitive data]
	[OCBUJ] 16:37:18 INFO: Starting web GUI on http://127.0.0.1:8080/
	[OCBUJ] 16:37:18 INFO: Starting UPnP discovery...
	[OCBUJ] 16:37:21 INFO: dial tcp: missing port in address 192.168.0.1
	[OCBUJ] 16:37:24 INFO: UPnP discovery complete (found 0 devices).
	[OCBUJ] 16:37:24 INFO: Starting local discovery announcements
	[OCBUJ] 16:37:24 INFO: Starting global discovery announcements
	[OCBUJ] 16:37:24 OK: Ready to synchronize [sensitive data] (read-write)
	[OCBUJ] 16:37:24 INFO: Device OCBUJGQ-[sensitive data] is "[sensitive data]" at [dynamic]
	[OCBUJ] 16:37:24 INFO: Device ZZMHPOP-[sensitive data] is "[sensitive data]" at [dynamic]
	[OCBUJ] 16:37:24 INFO: Device 4B7NWP2-[sensitive data] is "[sensitive data]" at [dynamic]
	[OCBUJ] 16:37:24 INFO: Starting usage reporting
	[OCBUJ] 16:37:24 INFO: Completed initial scan (rw) of folder [sensitive data]
	[OCBUJ] 16:38:00 INFO: Established secure connection to 4B7NWP2-[sensitive data] at 192.168.0.187:43360-109.[sensitive data]:13911
	[OCBUJ] 16:38:00 INFO: Device 4B7NWP2-[sensitive data] client is "syncthing v0.10.30"
	[OCBUJ] 16:38:00 INFO: Device 4B7NWP2-[sensitive data] name is "[sensitive data]"
	[OCBUJ] 16:38:02 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:03 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:04 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:05 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:06 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:07 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:08 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:09 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:11 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:12 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:14 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:38:14 WARNING: Folder "[sensitive data]" isn't making progress - check logs for possible root cause. Pausing puller for 1m0s.
	[OCBUJ] 16:39:15 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:16 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:17 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:18 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:19 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:20 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:21 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:22 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:23 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:24 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:25 INFO: Puller (folder "[sensitive data]", file "Unfa/unfa-01-23.wav"): pull: no available source device
	[OCBUJ] 16:39:26 WARNING: Folder "[sensitive data]" isn't making progress - check logs for possible root cause. Pausing puller for 1m0s.

### PR DESCRIPTION
**Pull request by bt90**
### Purpose

Fixes [#1628](https://github.com/syncthing/syncthing-android/issues/1628)

The basic idea is to continue with local discovery even if we can't list interfaces. Since Android doesn't allow us to collect our own device IPs, we let the receiving clients do the job by advertising unspecified addresses, as we do for global discovery.

Note that this only fixes IPv4 broadcast discovery. IPv6 multicast discovery relies on network interface information and is still broken.

### Testing

Tested on my Pixel 4a 5G. My other client running an unpatched syncthing v1.24.0 successfully resolved the unspecified addresses.

## Authorship

Your name and email will be added automatically to the AUTHORS file
based on the commit metadata.

### CODE REVIEW

File: lib/beacon/broadcast.go [hunk 1]
```diff
@@ -47,7 +47,6 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 		intfs, err := net.Interfaces()
 		if err != nil {
 			l.Debugln(err)
-			return err
```
Comment (bt90) [inline lib/beacon/broadcast.go:50]: `net.Interfaces()` will throw an error on Android. We can still carry on and use the fallback broadcast address `255.255.255.255` later on.

File: lib/beacon/broadcast.go [hunk 2]
```diff
@@ -59,7 +58,7 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 			addrs, err := intf.Addrs()
 			if err != nil {
 				l.Debugln(err)
-				return err
+				continue
```
Comment (bt90) [inline lib/beacon/broadcast.go:61]: Interface discovery might work while retrieving the addresses doesn't. So log the error and carry on.

File: lib/beacon/multicast.go [hunk 3]
```diff
@@ -135,6 +134,17 @@ func readMulticasts(ctx context.Context, outbox chan<- recv, addr string) error
 		joined++
 	}
 
+	// Handle failed interface lookup on Android
+	if len(intfs) == 0 {
+		err := pconn.JoinGroup(nil, &net.UDPAddr{IP: gaddr.IP})
```
Comment (bt90) [inline lib/beacon/multicast.go:139]: > JoinGroup uses the system assigned multicast interface when ifi is nil

https://pkg.go.dev/golang.org/x/net/ipv6#PacketConn.JoinGroup

File: lib/discover/local.go [hunk 4]
```diff
@@ -292,9 +289,9 @@ func (c *localClient) registerDevice(src net.Addr, device Announce) bool {
 	return isNewDevice
 }
 
-// filterUnspecifiedLocal returns the list of addresses after removing any
-// unspecified, localhost, multicast, broadcast or port-zero addresses.
-func filterUnspecifiedLocal(addrs []string) []string {
+// filterUndialableLocal returns the list of addresses after removing any
+// localhost, multicast, broadcast or port-zero addresses.
+func filterUndialableLocal(addrs []string) []string {
```
Comment (bt90) [inline lib/discover/local.go:294]: Renamed as we want to keep unspecified addresses, but still want to filter the remaining garbage.

**Comment by bt90** [comment]
Funnily enough, a potential fix for Golang itself seems to be emerging: https://github.com/golang/go/pull/61089

The question is whether it would be accepted and in which go version it would be delivered. The earliest date for a version with the fix would be February 2024 with go v1.22.

Should I continue with my tests, or is this too much of a hack and we'd better wait for a fix at the language level?

**Comment by imsodin** [comment]
> [...] we let the receiving clients do the job by advertising unspecified addresses, as we do for global discovery.

I didn't find where wo do that, i.e. where the receiver of broad-/multicast messages resolves the sender address when the advertised one is unspecified?

**Comment by bt90** [comment]
https://github.com/syncthing/syncthing/blob/f25a169c4c0943b34777a306e93d8fbefc9d8109/lib/discover/local.go#L232

**Comment by bt90** [comment]
The logging is a bit borked, but the IPv4 broadcast is properly transmitted and interpreted by my other client:

```
[NTMID] 2023/09/10 18:30:46.215548 local.go:206: DEBUG: discover: Received local announcement from 192.168.178.22:45269 for CG4YCIW
[NTMID] 2023/09/10 18:30:46.215548 local.go:235: DEBUG: discover: Registering addresses for CG4YCIW
[NTMID] 2023/09/10 18:30:46.215548 local.go:269: DEBUG: discover: Reconstructed URL is &url.URL{Scheme:"tcp", Opaque:"", User:(*url.Userinfo)(nil), Host:"192.168.178.22:22000", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
[NTMID] 2023/09/10 18:30:46.215548 local.go:271: DEBUG: discover: Replaced address 0.0.0.0 in tcp://0.0.0.0:22000 to get tcp://192.168.178.22:22000
[NTMID] 2023/09/10 18:30:46.215548 local.go:274: DEBUG: discover: Accepted address relay://88.212.32.226:22067/?id=5YBUXQ6 verbatim
[NTMID] 2023/09/10 18:30:46.215548 local.go:269: DEBUG: discover: Reconstructed URL is &url.URL{Scheme:"quic", Opaque:"", User:(*url.Userinfo)(nil), Host:"192.168.178.22:22000", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
[NTMID] 2023/09/10 18:30:46.215548 local.go:271: DEBUG: discover: Replaced address 0.0.0.0 in quic://0.0.0.0:22000 to get quic://192.168.178.22:22000
[NTMID] 2023/09/10 18:30:46.215548 local.go:274: DEBUG: discover: Accepted address quic://80.xxx.xxx.xxx:61166 verbatim
```

File: lib/beacon/broadcast.go (hunk 1 above)
Comment (acolomb) [reply lib/beacon/broadcast.go:50]: This comment should maybe live directly in the code.

File: lib/beacon/broadcast.go (hunk 2 above)
Comment (acolomb) [reply lib/beacon/broadcast.go:61]: Code comment instead of only GitHub?

**Comment by bt90** [comment]
Local discovery works over IPv4. Multicast doesn't seem to work unless an interface is explicitly selected. There are no errors logged other than the usual suspects, but I don't see any announcements from my local client.

I also noticed something else while testing this PR. Since interface listing doesn't work, our LAN detection on Android treats everything as WAN. So the address selection is a bit erratic.

**Comment by bt90** [comment]
I reverted the multicast changes as I couldn't get it to work. The local discovery via IPv4 broadcast is fixed though.

File: lib/beacon/broadcast.go [hunk 5]
```diff
@@ -46,8 +46,9 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 
 		intfs, err := net.Interfaces()
 		if err != nil {
-			l.Debugln(err)
-			return err
+			l.Debugln("Failed to list interfaces", err)
```
Comment (calmh) [inline lib/beacon/broadcast.go:49]:
```suggestion
			l.Debugln("Failed to list interfaces:", err)
```

File: lib/beacon/broadcast.go [hunk 6]
```diff
@@ -58,8 +59,9 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 
 			addrs, err := intf.Addrs()
 			if err != nil {
-				l.Debugln(err)
-				return err
+				l.Debugln("Failed to list interface addresses", err)
```
Comment (calmh) [inline lib/beacon/broadcast.go:62]:
```suggestion
				l.Debugln("Failed to list interface addresses:", err)
```

File: lib/discover/local.go [hunk 7]
```diff
@@ -310,8 +307,7 @@ func filterUnspecifiedLocal(addrs []string) []string {
 		switch {
 		case len(tcpAddr.IP) == 0:
 		case tcpAddr.Port == 0:
-		case tcpAddr.IP.IsUnspecified():
-		case !tcpAddr.IP.IsGlobalUnicast() && !tcpAddr.IP.IsLinkLocalUnicast():
+		case !tcpAddr.IP.IsGlobalUnicast() && !tcpAddr.IP.IsLinkLocalUnicast() && !tcpAddr.IP.IsUnspecified():
 		default:
```
Comment (calmh) [inline lib/discover/local.go:311]:
```suggestion
		case tcpAddr.IP.IsGlobalUnicast(), tcpAddr.IP.IsLinkLocalUnicast(), tcpAddr.IP.IsUnspecified():
```

let's use a positive expression instead

**Review by calmh — APPROVED** [review]

**Comment by calmh** [comment]
Wow, so the "Fixes" in the body there becomes very misleading with GitHub rendering... For the record, we do those things in the commit subject, and none of the tooling supports linking in issues from other repos... 🤷

**Comment by tomasz1986** [comment]
> Wow, so the "Fixes" in the body there becomes very misleading with GitHub rendering... For the record, we do those things in the commit subject, and none of the tooling supports linking in issues from other repos... 🤷

I think I've seen this in other issues too. GitHub really should display the link as `syncthing-android#1628` or similar to avoid the confusion.

**Pull request merged**